update guidelines (#43)
openalea/openalea.rtfd.io: doc/development/guidelines.md

  - Compatible `Python` version: [![Python Version](https://img.shields.io/badge/python-3.8%20%7C%203.9%20%7C%203.10%20%7C%203.11%20%7C%203.12-blue)](https://www.python.org/downloads/)
  - License: [![License](https://img.shields.io/badge/License--CeCILL-C-blue)](https://www.cecill.info/licences/Licence_CeCILL-C_V1-en.html)
  - Version of the package on Anaconda: [![Anaconda-Server Badge](https://anaconda.org/openalea3/mtg/badges/version.svg)](https://anaconda.org/openalea3/mtg)

> - installation instructions: how to install the package, for user (using `conda`) and for developer (using conda and `pip -e`), e.g. :

Reviewer: ```suggestion
- installation instructions: how to install the package for user (using `conda` / `mamba`) and for developer (using `conda` / `mamba` and `pip -e`). For more information about how to declare package dependencies for usage and development, cf. [pyproject.toml](#pyprojecttoml) and [build the package](#building-the-package) sections. 
```

from importlib.metadata import version, PackageNotFoundError

try:
    __version__ = version("openalea.my_pkg")
except PackageNotFoundError:
    # package is not installed
    pass
```
- To avoid confusion all other manual reference to version (version.py files,...) should be removed from your source tree

## pyproject.toml

OpenAlea projects should use the `pyproject.toml` file to define the build configuration and the metadata of the project.
This file should contain the following sections:

- In most cases you can use [`setuptools` as the build system](https://setuptools.pypa.io/en/latest/). `setuptools` is the most widely used build system for Python packages. It is used to define the metadata of the package, the dependencies, and the entry points.
We also recommend using setuptools_scm as a companion tool to handle automatically set the version metadata of your package based on git tags.
```toml
[build-system]
requires = [
    "setuptools", 
    "setuptools_scm",
]
build-backend = "setuptools.build_meta"

# allow openalea to be a namespace package
[tool.setuptools.packages.find]
where = ["src"]

# enable dynamic versioning based on git tags

> [tool.setuptools_scm]

Author: idem as above : we do not need to set options, but we need to activate it by setting this line

  "pytest >=6",
  "pytest-cov >=3",
]
doc = [
  "pydata-sphinx-theme",
  "myst-parser",
  "sphinx-favicon",
  "ipykernel",
  "sphinx-copybutton",
  "ipython_genutils",
  "nbsphinx",
]
```

- Define the URLs of the project in the `[project.urls]` section.

```toml
[project.urls]
Repository = "https://github.com/openalea/pkg_name"

> Homepage = "https://openalea.readthedocs.io/"

Author: I had to use 'openalea-pkg_name'  for astk as astk.rtfd.io already exist. But for most package pkg_name.rtfd.io is available and could be used. => modify guidelines accordingly !

- all dependence should also be available from a conda channel or via `pip`.
- package should be built and uploaded to the `openalea3` conda channel. A [dedicated CI/CD pipeline](#ci-cd) can be used for this purpose.
- use GIT_DESCRIBE_TAG conda variable to automatically set conda version to your last tag

We also recommend re-using the informations declared in your pyproject.toml to source only once your list of dependencies.

A minimal conda build information could be provided by adding the following generic conda/meta.yaml file at the root of your project:

```yaml
{% set pyproject = load_file_data('../pyproject.toml', from_recipe_dir=True) %}

package:
  name: {{ pyproject["project"]["name"] }}
  version: {{ GIT_DESCRIBE_TAG }}

source:
  git_url: {{ pyproject["project"]["urls"]["Repository"] }}

build:
  noarch: python
  preserve_egg_dir: True
  script: {{ PYTHON }} -m pip install . -vv

> requirements:

Reviewer 2: not sure this is best way. Some packages have not the same names btw conda and pip

Reviewer: Not sure to understand your comment @pradal. Are you refering the `package.name` line ? Is this not what we want ?

Reviewer 2: Requirements are used by pip. But some packages are not release with the same name in pip and conda.
And sometimes explicit is better than implicit. In meta.yaml, I will prefer to explicitly define the deps.
And to not install deps with pip but only with conda.

Reviewer: > Requirements are used by pip. But some packages are not release with the same name in pip and conda. And sometimes explicit is better than implicit. In meta.yaml, I will prefer to explicitly define the deps. 

This is not in line with our will to indicate everything in one place / file, a.k.a. `pyproject.toml` though...

In our case, using [pixi](https://pixi.sh/latest/) would help a lot: it would let use define conda dependencies by default in the `pyproject.toml` and build more easily, even though this adds another tool and dependancy. I will give it a try on a simple openalea package...

Author: At conda-build stage, all requirements will be installed by conda (see meta.yaml).
Pip will be in charge of dependencies only for documentation (that uses environment.yml).
For packages that have different names in pip and conda, I think they should be considered as package only available on conda. I may better precise that in the doc?

Author: Okay I add that in toml documentation in the guidelines:

# you can list here all dependencies taht are pip-intalable, and that have a name understood by conda (to allow reuse of this list in meta.yaml)
# If conda name is different, please do not declare the pip name, and declare in the next section the conda name
dependencies = [
    "numpy >= 1.24",
    ...
]

# section specific to conda-only distributed package (not used by pip yet)
[tool.conda.environment]
channels = [
  "openalea3",
  "conda-forge"
]
dependencies = [
    "openalea.plantgl",
    "openalea.deploy",
    ...
]

Is it okay ?

  build:
    - python
    {% for dep in pyproject["build-system"]["requires"] %}
    - {{ dep.lower() }}
    {% endfor %}

  run:
    - python {{ pyproject["project"]["requires-python"] }}
    {% for dep in pyproject["project"]["dependencies"] %}
    - {{ dep.lower() }}
    {% endfor %}

test:
  requires:
    - pytest
  imports:
    - {{ pyproject["project"]["name"] }}
  source_files:
    - test/test_*.py
  commands:
   - cd test
   - pytest -v
```

- Using git_url in source is mandatory to have GIT_DESCRIBE_TAG correctly set during conda build on CI.
However, when testing locally the build of your package, it will use the version pushed on git. 
To shorten the debugging cycle and directly see the effect of uncommited modifications, 'source' can be temporarilly set to (but not commited !):

```yaml
source:
  path: ../
```

- You can also provide a conda/environment.yml file that will ease a developper installing in develop mode:
- 
```yaml
name: mypkg_dev
channels:

>   - conda-forge

Reviewer 2: I will recommend to add openalea3 before conda-forge

Reviewer: At least need to be consistent with the way it is defined in the `pyproject.toml`. No possibility to retrieve that by the way ? 
```suggestion
channels:
  {% for channel in pyproject["project"]["tool.conda.environment"]["channels"] %}
  - {{ channel.lower() }}
  {% endfor %}
```

Author: Aggre with both in theory : invert the order in pyproject.toml and add the re-use in environement.yml. I am just not sure that load_file_data is available outside meta.yaml

Author: I checked that pyproject can be loaded in meta.yaml only. The only way to re-use declaration in pyproject dependencies is therefore to let pip install the pip-instalable dependencies...at the cost of a conda-pip mix environment...that generaly will only be used to test/build the doc.

  - openalea3
dependencies:
  - python
  - pip
  - pandoc
# list here conda-only deps
  - openalea.plantgl
  - pip:

>       - -e ..[doc,test]

Reviewer 2: pip -e is mainly for doc. Not for a conda environment.yaml

Reviewer: - Why do we need the editable install, why not the "regular" install ?
- My understanding is that the `optional-dependencies` defined in the `pyproject.toml` are needed for developpers to test / document / visualize / etc... So does this need to be in the conda package ?

Author: environment.yml is not used for conda package builds. I agree that optional dependencies are not needed to be included: that's why I do not include them in meta.yaml. However if I define 'conda only' but required dependencies in pyproject.toml tools.conda.environment, They should be read and added in meta.yaml using a jinja 'for loop'.
Again this environment is precisely for doc and optionally for developpers that are okay with the idea of mixing pip/conda. To me its main interest is to allow a fast and easy isolated working environment setup  for testing/building doc/package from scratch in a astk-dedicated environment.
For other cases (eg developping astk functionalities together with adel and caribu) I will not use that file, but conda --only_deps intall type (see astk.readme)

Reviewer: Ok for me.

Shouldn't it be worth adding a dedicated section in this page `Create a developer's environment` with this info, and the different use cases you mention, e.g. as done in [astk](https://github.com/openalea-incubator/astk/tree/master?tab=readme-ov-file#developpers)

Author: I have add a section detailing that as an example of the installation instruction in the 'README.md' desction of the guidelines. Is it okay for you  #?

Reviewer: OK. I suggest adding the links / references to dedicated sections in the `README.md` section of the guidelines.

Author: Okay ! Thanks for the suggestions/modifications !

```yaml
{% set pyproject = load_file_data('../pyproject.toml', from_recipe_dir=True) %}

package:
  name: {{ pyproject["project"]["name"] }}
  version: {{ GIT_DESCRIBE_TAG }}

source:
  path: ..

build:
  noarch: python
  preserve_egg_dir: True
  script: {{ PYTHON }} -m pip install . -vv

requirements:
  build:
    - python
    {% for dep in pyproject["build-system"]["requires"] %}
    - {{ dep.lower() }}
    {% endfor %}

  run:
    - python {{ pyproject["project"]["requires-python"] }}
    {% for dep in pyproject["project"]["dependencies"] %}
    - {{ dep.lower() }}
    {% endfor %}

test:
  requires:
    - pytest
  imports:
    - {{ pyproject["project"]["name"] }}
  source_files:
    - test/test_*.py
  commands:
   - cd test
   - pytest -v

about:
  home: {{ pyproject["project"]["urls"]["Homepage"] }}
  license: {{ pyproject["project"]["license"] }}
  summary: {{ pyproject["project"]["description"] }}
```

- You can also provide a conda/environment.yml file that will ease a developer installing in develop mode:
- 
```yaml
name: mypkg_dev
channels:
  - conda-forge
  - openalea3
dependencies:
  - python
  - pip
  - pandoc
# list here conda-only deps
  - openalea.plantgl

>   - pip:

Reviewer: But is it even ok to have pip dependencies in a conda package ?

>   - pip:

Author: Here were are not in conda-build : the package built by conda will be only with conda deps.
This environment is for readthedoc/developpers that want to install from scratch. In astk/readme, I also give a line for developpers who want to install using conda only and no mix

>   - pip:

Reviewer: Same comment as above now I got it that we're talking about the `environment.yaml` file. Ok for me with the "warning" that we're mixing conda and pip installs. In the past this used to be a concern, but from what I've tested recently, this is not really the case anymore.

>   - pip:

Author: I update with:
- You can also provide a conda/environment.yml file that will ease maintainers developing in a isolated environment, and also be used by readthedoc):
- 
```yaml
name: mypkg_dev
channels:
  - openalea3
  - conda-forge
dependencies:
  - python
  - pip
  - pandoc
# list here manually conda-only deps (listed in [tool.conda.environment] section of pyproject)
  - openalea.plantgl
# let pip install the rest using pyproject.toml (if you are okay with conda/pip mix)
  - pip:
      - -e ..[doc,test]
```
Is it okay ?

A minimal conda build information could be provided by adding the following generic conda/meta.yaml file at the root of your project:

```yaml
{% set pyproject = load_file_data('../pyproject.toml', from_recipe_dir=True) %}

package:
  name: {{ pyproject["project"]["name"] }}
  version: {{ GIT_DESCRIBE_TAG  | replace("v", "") }}

source:
  path: ..

build:
  noarch: python
  preserve_egg_dir: True
  script: {{ PYTHON }} -m pip install . -vv

requirements:
  build:
    - python
    {% for dep in pyproject["build-system"]["requires"] %}
    - {{ dep.lower() }}
    {% endfor %}

  run:
    - python {{ pyproject["project"]["requires-python"] }}
    {% for dep in pyproject["project"]["dependencies"] %}
    - {{ dep.lower() }}
    {% endfor %}

test:
  requires:
    - pytest
  imports:
    - {{ pyproject["project"]["name"] }}
  source_files:
    - test/test_*.py
  commands:
   - cd test
   - pytest -v

about:
  home: {{ pyproject["project"]["urls"]["Homepage"] }}
  license: {{ pyproject["project"]["license"] }}
  summary: {{ pyproject["project"]["description"] }}
```

> - You can also provide a conda/environment.yml file that will ease maintainers developing in a isolated environment, and also be used by readthedoc):

Reviewer: ```suggestion
- You can also provide a conda/environment.yml file that will ease maintainers developing in a isolated environment (and can also be used by readthedoc):
```

      anaconda_token: ${{ secrets.ANACONDA_TOKEN }}
```

> This action will build the package on a matrix of operating systems (`[ubuntu-latest , macos-latest , windows-latest]`) and Python versions (`[3.8, 3.9, 3.10, 3.11, 3.12]`) every time a new commit is pushed.

Reviewer: ```suggestion
This action will build the package on a matrix of operating systems (`[ubuntu-latest , macos-latest , windows-latest]`) and Python versions (`[3.9, 3.10, 3.11, 3.12, 3.13]`) every time a new commit is pushed.
```

      anaconda_token: ${{ secrets.ANACONDA_TOKEN }}
```

This action will build the package on a matrix of operating systems (`[ubuntu-latest , macos-latest , windows-latest]`) and Python versions (`[3.8, 3.9, 3.10, 3.11, 3.12]`) every time a new commit is pushed.

To enable a new upload to openalea3 conda channel, just create a tag/release starting with `v` on your github master. 
To summarize we recommend the following development workflow: create branches, make pull request, review them, merge into master, and create a new tag release from github web interface.If you followed the guideline above, the tag wll be propagated to the package metadata and to the conda package.

## Documentation

The documentation of the package should be written using the [Sphinx](https://www.sphinx-doc.org/en/master/) documentation generator and hosted on the [ReadTheDocs](https://readthedocs.org/) platform.

> To set up the ReadTheDocs documentation, you need to create a ReadTheDocs account and add a `.readthedocs.yml` file to the root of your project with the following content:

Reviewer: Do we really need to create an account though ? If the project is in `openalea` or `openalea-incubator`, I think this is not needed.

Reviewer 2: Yes. And mainly we want to have access to all these packages from openalea devs.
So I will remove these lines.

Author: Okay, I did the same ! I will add 'log in with your github account' instead of 'create an account' !

Reviewer 2: No. Just say log in on rtfd.
How each user do that is the responsability of rtfd, not our.

```

> This file will tell ReadTheDocs to build the documentation using the environment describe in the `doc/environment.yml` file (that can be identical to conda/environment.yml) and to set up the environment using `mambaforge`.

Reviewer: ```suggestion
This file will tell ReadTheDocs to build the documentation using the environment describe in the `doc/environment.yml` file (that does not need to be identical to conda/environment.yml) and to set up the environment using `mambaforge`.
```

Author: ```suggestion
This file will tell ReadTheDocs to build the documentation using the environment describe in the `doc/environment.yml` file (you can also use conda/environment.yml) and to set up the environment using `mambaforge`.
```
